File: Server/src/resource-node.ts
import { Config } from "./config.js";
import { ResourceType, type ResourceNodeState } from "./types.js";

const RESOURCE_TYPES = Object.values(ResourceType);

let nextNodeId = 1;

export class ResourceNode {
  readonly id: number;
  readonly type: ResourceType;
  x: number;
  y: number;
  totalAmount: number;
  currentAmount: number;
  quality: number; // 1-5
  respawnTime: number; // ms
  private depletedAt: number | null = null;

  constructor(
    type: ResourceType,
    x: number,
    y: number,
    quality: number,
    totalAmount?: number,
  ) {
    this.id = nextNodeId++;
    this.type = type;
    this.x = x;
    this.y = y;
    this.quality = Math.max(1, Math.min(5, Math.round(quality)));
    this.totalAmount =
      totalAmount ??
      Math.round(
        Config.nodeBaseAmount *
          (1 + (this.quality - 1) * Config.nodeQualityMultiplier),
      );
    this.currentAmount = this.totalAmount;
    this.respawnTime = Config.nodeRespawnTime;
  }

  /** Returns true when the node has resources available. */
  get isAvailable(): boolean {
    return this.currentAmount > 0;
  }

  /** Returns true when the node is depleted and waiting to respawn. */
  get isDepleted(): boolean {
    return this.depletedAt !== null;
  }

  /**
   * Extract resources from this node.
   * Returns the actual amount extracted (may be less than requested if near depletion).
   */
  extract(amount: number): number {
    if (!this.isAvailable) return 0;
    const actual = Math.min(amount, this.currentAmount);
    this.currentAmount -= actual;
    if (this.currentAmount <= 0) {
      this.currentAmount = 0;
      this.depletedAt = Date.now();
    }
    return actual;
  }

  /** Called each tick; returns true if the node just respawned. */
  tickRespawn(): boolean {
    if (this.depletedAt === null) return false;
    if (Date.now() - this.depletedAt >= this.respawnTime) {
      this.currentAmount = this.totalAmount;
      this.depletedAt = null;
      return true;
    }
    return false;
  }

  /** Serialize to a payload-friendly shape. */
  toState(): ResourceNodeState {
    return {
      id: this.id,
      type: this.type,
      x: this.x,
      y: this.y,
      currentAmount: this.currentAmount,
      totalAmount: this.totalAmount,
      quality: this.quality,
    };
  }
}

/** Spawn a set of resource nodes randomly within the given bounds. */
export function spawnResourceNodes(
  count: number,
  bounds: { xMin: number; xMax: number; yMin: number; yMax: number },
): ResourceNode[] {
  const nodes: ResourceNode[] = [];
  for (let i = 0; i < count; i++) {
    const type = RESOURCE_TYPES[Math.floor(Math.random() * RESOURCE_TYPES.length)]!;
    const x = bounds.xMin + Math.random() * (bounds.xMax - bounds.xMin);
    const y = bounds.yMin + Math.random() * (bounds.yMax - bounds.yMin);
    const quality = Math.ceil(Math.random() * 5);
    nodes.push(new ResourceNode(type, x, y, quality));
  }
  return nodes;
}

File: ServerRS/src/game/resource_node.rs
use std::sync::atomic::{AtomicU32, Ordering};
use std::time::Instant;

use rand::Rng;

use crate::config;
use crate::msg::types::ResourceNodeState;

use super::zone::ZoneBounds;

static NEXT_NODE_ID: AtomicU32 = AtomicU32::new(1);

const RESOURCE_TYPES: &[&str] = &["iron", "copper", "silicon", "carbon", "titanium"];

pub struct ResourceNode {
    pub id: u32,
    pub resource_type: String,
    pub x: f64,
    pub y: f64,
    pub total_amount: f64,
    pub current_amount: f64,
    pub quality: u32, // 1-5
    respawn_time_ms: u64,
    depleted_at: Option<Instant>,
}

impl ResourceNode {
    pub fn new(resource_type: String, x: f64, y: f64, quality: u32) -> Self {
        let quality = quality.clamp(1, 5);
        let total_amount = (config::NODE_BASE_AMOUNT
            * (1.0 + (quality as f64 - 1.0) * config::NODE_QUALITY_MULTIPLIER))
            .round();

        Self {
            id: NEXT_NODE_ID.fetch_add(1, Ordering::Relaxed),
            resource_type,
            x,
            y,
            total_amount,
            current_amount: total_amount,
            quality,
            respawn_time_ms: config::NODE_RESPAWN_TIME_MS,
            depleted_at: None,
        }
    }

    pub fn is_available(&self) -> bool {
        self.current_amount > 0.0
    }

    /// Extract resources. Returns actual amount extracted.
    pub fn extract(&mut self, amount: f64) -> f64 {
        if !self.is_available() {
            return 0.0;
        }
        let actual = amount.min(self.current_amount);
        self.current_amount -= actual;
        if self.current_amount <= 0.0 {
            self.current_amount = 0.0;
            self.depleted_at = Some(Instant::now());
        }
        actual
    }

    /// Called each tick; returns true if the node just respawned.
    pub fn tick_respawn(&mut self) -> bool {
        if let Some(depleted_at) = self.depleted_at {
            if depleted_at.elapsed().as_millis() as u64 >= self.respawn_time_ms {
                self.current_amount = self.total_amount;
                self.depleted_at = None;
                return true;
            }
        }
        false
    }

    pub fn to_state(&self) -> ResourceNodeState {
        ResourceNodeState {
            id: self.id,
            node_type: self.resource_type.clone(),
            x: self.x,
            y: self.y,
            current_amount: self.current_amount,
            total_amount: self.total_amount,
            quality: self.quality,
        }
    }
}

pub fn spawn_resource_nodes(count: usize, bounds: &ZoneBounds) -> Vec<ResourceNode> {
    let mut rng = rand::thread_rng();
    let mut nodes = Vec::with_capacity(count);

    for _ in 0..count {
        let resource_type =
            RESOURCE_TYPES[rng.gen_range(0..RESOURCE_TYPES.len())].to_string();
        let x = rng.gen_range(bounds.x_min..bounds.x_max);
        let y = rng.gen_range(bounds.y_min..bounds.y_max);
        let quality = rng.gen_range(1..=5);
        nodes.push(ResourceNode::new(resource_type, x, y, quality));
    }

    nodes
}
